### PR TITLE
Pass compilation context via swift info in hmap creation due to rules_swift 2.1.1

### PR DESCRIPTION
**Pull request by gyfelton**
One of the CI error is:
```
error: generate-pch command failed with exit code 1 (use -v to see invocation)
tests/ios/unit-test/test-imports-app/TestImports-App-Bridging-Header.h:1:9: error: 'TestImports-App/Header2.h' file not found
#import <TestImports-App/Header2.h>
        ^
1 error generated.
<unknown>:0: error: failed to emit precompiled header 'bazel-out/ios-sim_arm64-min12.0-applebin_ios-ios_sim_arm64-dbg-ST-c2aefc9133a8/bin/_pch_output_dir/TestImports-App-Bridging-Header-swift_28IU2BV1DK30K-clang_1FNSWCOAS4SUQ.pch' for bridging header 'tests/ios/unit-test/test-imports-app/TestImports-App-Bridging-Header.h'
```
This is because this header was suppose to be an input to Swift compile via hmap creation rule. But in this breaking change https://github.com/bazelbuild/rules_swift/commit/d68b21471e4e9d922b75e2b0621082b8ce017d11 rules_swift stopped collecting compilation context from `CcInfo` if the target is Swift code (line 1555 of `swift/internal/compiling.bzl`). Instead it depends on reading the same info from `clang_module.compilation_context` which is available from `SwiftInfo` provider (SwiftInfo -> modules -> clang)

This PR deals with header maps side where we propagate compilation context via this `clang` module

Next step is to do the same for other places where we need to pass the compilation context 

CI for this PR should no longer produce this error (but will still fail on some other tests)

### CODE REVIEW

File: rules/hmap.bzl
Comment (congt) [inline rules/hmap.bzl:95]: Is cc_info_provider still necessary for rules_swift 1.18.0?

Comment (gyfelton) [reply rules/hmap.bzl:95]: 1.18 def. need it
not sure about 2.1.1, so far it seems that the logic depends on it if it is not a swift file to compile?
based on the commit i think it is `apple_common.new_objc_provider()` we will remove soon as latest bazel don't have Objc provider under apple_common anymore: https://bazel.build/rules/lib/toplevel/apple_common (7.3 still has it)

**Review by luispadron — APPROVED** [review]
Could we stack this on #903 just to confirm CI is all green before merging both

**Comment by gyfelton** [comment]
> Could we stack this on #903 just to confirm CI is all green before merging both

changed base to luis/add-clang-attr-to-merged-swift-info-for-framework and will merge that branch once the PR is all green

**Pull request merged**
